### PR TITLE
Add some code gen stuff

### PR DESCRIPTION
**Pull request by duglin**
Add some code gen stuff - based off of @jpeeler work

This will not fully build w/o https://github.com/kubernetes-incubator/service-catalog/pull/187

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

**Comment by MHBauer** [comment]
is it possible to `go get k8s.io/kubernetes/cmd/libs/go2idl/deepcopy-gen` & `go get k8s.io/kubernetes/cmd/libs/go2idl/defaulter-gen` or are we customizing this code somehow that we need to include it in our repo?

**Comment by duglin** [comment]
@MHBauer probably. For my testing I was making mods to it just for debugging, but once we're past that I think we can probably do what you're suggesting.  Although, it wants to put it in MY "bin" dir rather than our src tree's bin dir. Any one know if its possible to keep our GOPATH as it it now but have "go get" put the exe into a different path?

**Comment by duglin** [comment]
oh wait, I forgot that we do have our own main.go for each - you'd have to ask @jpeeler what special mods he may have made in those.

**Comment by pmorie** [comment]
@duglin @MHBauer 

> is it possible to go get k8s.io/kubernetes/cmd/libs/go2idl/deepcopy-gen & go get k8s.io/kubernetes/cmd/libs/go2idl/defaulter-gen or are we customizing this code somehow that we need to include it in our repo?

Right now those binaries have hard-coded package names in them, so we'll need the versions @jpeeler put together for it to work.  Long-term the binaries should be go-gettable and usable in any project, but they aren't at the moment.

File: Makefile [hunk 1]
```diff
@@ -28,7 +28,7 @@ endif
 # Some will have dedicated targets to make it easier to type, for example
 # "apiserver" instead of "bin/apiserver".
 #########################################################################
-build: init \
+build: init .generate_files \
```
Comment (pmorie) [inline Makefile]: I don't want to regenerate the files each time I build.  It takes too long.

File: Makefile [hunk 2]
```diff
@@ -54,6 +54,28 @@ apiserver: $(BINDIR)/apiserver
 $(BINDIR)/apiserver: cmd/service-catalog
 	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
 
+# This section contains the code generation stuff
+#################################################
+.generate_exes: $(BINDIR)/defaulter-gen $(BINDIR)/deepcopy-gen
```
Comment (pmorie) [inline Makefile]: I'm going to need someone more familiar with make to explain this to me.

File: Makefile (hunk 1 above)
Comment (duglin) [inline Makefile]: it doesn't

File: Makefile (hunk 2 above)
Comment (duglin) [inline Makefile]: if the timestamp on the src files are newer than the  .generate_exes file then it'll rebuild.  Technically it should check the types.go/doc.go files too.

Comment (pmorie) [inline Makefile]: I guess it's time to really RTFM for me for make.

**Comment by pmorie** [comment]
After some thought, I think I would prefer to merge this over the stuff that @jpeeler opened in #180 **on the condition that it generates the same output**.  @duglin can you verify that it generates the exact same output tomorrow?  I am suspicious that there is some finer point lost here because it's SO much more simple, but I'm actually finding myself more interested in getting the functionality without an exact repetition of what happens upstream than I anticipated that I would be.

Also, I am going to create an issue to make the generator binaries generic, so that they can be go-gettable and/or published in a docker image.  It looks like the only differentiator between ours and those in k8s is that the package is hard-coded.

File: Makefile (hunk 2 above)
Comment (duglin) [inline Makefile]: Nothing special here, just normal timestamp processing/checking.

**Comment by duglin** [comment]
#188 and #192 should probably go in first - to avoid the lint issues

**Comment by duglin** [comment]
rebased and all prereq PRs are in now.

File: .gitignore [hunk 3]
```diff
@@ -13,3 +13,5 @@ coverage.html
 .glide
 .scBuildImage
 .init
+.generate*
+**/zz*.go
```
Comment (arschles) [inline .gitignore]: We talked offline yesterday about checking in generated code, and I think we're in agreement that it's a good idea to check in. Can we remove this line to achieve that?

Comment (pmorie) [inline .gitignore]: I'm good with the rest of this PR, minus this line.

Comment (duglin) [inline .gitignore]: I didn't agree to that :-)
I'm still waiting for a good reason for us to check it in. Yes I can think of one or two but none of them apply to this project yet.  I see no reason not to generate this like any other output file during the build process.

People have expressed concerns in the past with checking-in vendored code because someone may "accidentally" modify it - how is this any different?

Comment (MHBauer) [inline .gitignore]: Why check it in?

Comment (thockin) [inline .gitignore]: some history and current state on generated code, from the peanut gallery.

Historically (in the main kube repo) we *had* to commit  generated code, because `go get` was important, because the only real API client was in the same repo.  Nowadays, the client snapshots all it needs into a new repo, so `go get` can take a long walk off a short pier (for the main repo).

Do you need this repo to be `go get` compatible?  If so, you pretty much have to check in the generated code.

The alternative to committing them is to have fairly complicated builds that auto-detect when to rebuild them.  The main repo has a (perhaps overly complicated) solution for this, wrt a handful of code-generators, but not all.  I see this PR has a less "sophisticated" build here, which seems OK at first blush.

So the main reason that zz_generated.* is checked into the main repo is that I have not resurrected the PR to remove them.  I will do once 1.5.x cherrypicks settle down.

So philosophically I *want* to un-commit generated code.  It's insane how much rebasing happens because of it.  It's not as simple (for the main repo).  If you need `go get` or if you start to do the JSON-generation or proto-generation stuff, you will get ahead of my work, and be forced to check it in (or solve the build yourself).  I'd be happy for help, but it's a massive distraction from "productive" work.

Looking at this PR, it's obvious that even the work we have done so far wrt deepcopy is not done enough.

TL;DR: no, go back and read what I wrote.  It's too complicated to summarize.

Comment (arschles) [inline .gitignore]: Sorry @duglin, I realize now that I assumed but didn't confirm that you agree to it.

Anyway, to recap, here are the reasons I gave on Slack yesterday, in favor of checking in generated code:

1. There's no dependency on code generators for devs (except those who want to make changes to the IDL file) and CI
2. We can build a reusable tiller SDK for 3rd parties (i.e. the original steward used that SDK)

Additional notes:

1. Accidental modification was one of the concerns with checking in vendored code
1. Generated code is not the same as vendored code. Its source and output are isolated to this repository
    1. If someone accidentally modifies generated code and reviewers don't catch the change, the build will fail next time code is generated. But most importantly, when the build fails, a simple diff in our codebase (not in a dependency) shows why. 
1. I must repeat the sentiment that it's important that we follow the general structure of kubernetes/kubernetes for now as best we can. This is to make it more friendly to core developers who may want to contribute to, extend, or otherwise adopt our codebase -- a goal of all of ours

Comment (duglin) [inline .gitignore]: None of the reasons for us *having* to check in our code apply to our repo yet - as far as I can tell. And I'm all for checking them in when we *have to*.  But, given @thockin's statement about trying to remove it from the main repo, it seems to me that it might actually be a benefit for us to be a guinea pig in this area - let us be the a testbed for the rest of Kube because if we find that we can't simplify the process then it may not be worth the effort for the main repo to try.  On the other hand, perhaps something can be learned from our trailblazing here - perhaps not the dream goal of not checking it in at all, but maybe we'll land on a middle group position. I would just prefer to not toss in the towel w/o trying.

**Comment by thockin** [comment]
NB: that your build (seems to?) regenerate for every `make`, which means
there will never be a do-nothing build.  If you're OK with the speed of
that, go nuts.  To fix it properly is VERY complex in Makefile.  Bazel may
obviate a lot of that complexity.  Want to be a guinea pig - make bazel
work :)

@mikedanese

**Comment by duglin** [comment]
@thockin hmmm I'm not seeing that. For me "make build" once done, does nothing on subsequent runs.   but let me try some more...

File: .gitignore (hunk 3 above)
Comment (pmorie) [inline .gitignore]: > I would just prefer to not toss in the towel w/o trying.

No one's tossing in any towel.  Check it in for now, remove it in a follow up.  We have git.

**Comment by pmorie** [comment]
I'm going to reiterate my position in the main thread:

1.  Check this code in for now
2.  In a follow-up, make it so it's not committed

Note, **this blocks me** from moving forward on other code generation activities.  I would really like to have this in, so that I can build in it with confidence, and get everything just-so later.  Let me also reiterate that we have agreed, as a group, not to let perfection be the enemy of progress.  There is literally *no reason* to avoid changing this PR so that it can be merged now, from what I can tell.

**Comment by krancour** [comment]
LGTM

**Comment by MHBauer** [comment]
lgtm as is.

**Comment by thockin** [comment]
regarding the build: Yes, you're right.  It won't rebuild on every make.
It also won't rebuild when the inputs change, which is arguably worse.  You
need to remember to `rm -f .generate_*` whenever you make a change to your
API files that might change the codegen outputs.

Make depends on comprehending a DAG, but you're not giving it a full DAG.
There's a reason the kube Makefiles are so horrific.

**Comment by duglin** [comment]
@thockin yup - I know the list of dependencies isn't right yet - but that's a WIP

**Comment by thockin** [comment]
Roger that.  I predict that if you continue to work it, you will arrive at
what we have in Kube Makefile :)  If you're going to spend energy on it,
I'd beg that you spend the energy to make Kube Makefile cleaner or more
reusable.

**Comment by pmorie** [comment]
In the name of expediency, I'm going to merge this.  I've created #203 to address the concern of which generated code is checked in in the long term.

**Comment by pmorie** [comment]
Part of #178
